sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,655 @@
+  auto ptr = ...
+  // alignment of the pointer in bytes specified using the property 'alignment'
+  auto arg_a = annotated_ptr(ptr, properties{alignment<4>});
+
+  q.submit([=]{
+    ...
+    *arg_a = (*arg_a) * 2;
+  });
+  ...
+}
+----
+
+The section below and the table that follows, describe the constructors, member
+functions and factory methods for `annotated_ptr`.
+The section below refers to an `annotated_ref` class which is described in the
+section following this one.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+template <typename T, typename PropertyListT = properties<>>
+class annotated_ptr {
+  public:
+    using reference = annotated_ref<T, PropertyListT>;
+
+    annotated_ptr() noexcept;
+    explicit annotated_ptr(T *Ptr, const properties<PropertyListT> &) noexcept;
+    annotated_ptr(annotated_ptr const &) noexcept;
+    template <T, typename P> explicit annotated_ptr(
+      annotated_ptr<T, P> const &) noexcept;
+    template <T, typename PropertyListU, typename PropertyListV>
+    explicit annotated_ptr(annotated_ptr<T, PropertyListU> const &,
+      properties<PropertyListV>) noexcept;
+
+    reference operator*() const noexcept;
+    reference operator[](std::ptrdiff_t) const noexcept;
+    annotated_ptr operator+(size_t) const noexcept;
+    std::ptrdiff_t operator-(annotated_ptr) const noexcept;
+
+    operator bool() const noexcept;
+
+    // Implicit conversion is not supported
+    operator T*() noexcept = delete;
+    operator const T*() const noexcept = delete;
+
+    T* get() noexcept;
+    const T* get() const noexcept;
+
+    annotated_ptr& operator=(const T*) noexcept;
+    annotated_ptr& operator=(annotated_ptr const&) noexcept;
+
+    annotated_ptr& operator++() noexcept;
+    annotated_ptr operator++(int) noexcept;
+    annotated_ptr& operator--() noexcept;
+    annotated_ptr operator--(int) noexcept;
+
+    template<typename propertyT>
+    static constexpr bool has_property();
+
+    // The return type is an unspecified internal class used to represent
+    // instances of propertyT
+    template<typename propertyT>
+    static constexpr /*unspecified*/ get_property();
+
+  private:
+    T *Ptr;
+  };
+} // namespace sycl::ext::oneapi
+
+----
+
+[frame="topbot",options="header"]
+|===
+|Functions |Description
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr() noexcept;
+----
+|
+Constructs an `annotated_ptr` object. Does not allocate new storage. The
+underlying pointer is initialized to `nullptr`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+explicit annotated_ptr(T *Ptr, const properties<PropertyListT> &P) noexcept;
+----
+|
+Constructs an `annotated_ptr` object. Does not allocate new storage. The
+underlying pointer is initialized with `Ptr`. `P` is used to specify the
+`PropertyListT` type on the class.
+
+The new property set `PropertyListT` must contain all properties from `P`,
+and if any common property takes a value, the value must be the same.
+
```

Comment:
Should P be optional?  That would allow for code like this to be written very succinctly (but I'm not sure how important this use case is):

```
template<typename props>
void foo(int *a, annotated_ptr<int, props> b) {
  // Create an annotated version of a with the same properties as b:
  decltype(b) annotated_a{a};
}
```

Reply:
I thought of a variant that might be more realistic: casting an annotated_ptr from one type to another while keeping the annotations:

```
template<typename src, typename dst, typename props>
annotated_ptr<dst, props> ptr_cast_while_preserving_annotations(annotated_ptr<src, props> src) {
  annotated_ptr<dst, props> temp{src.get()};
  return temp;
}
```